apps/wallet/src/background/keyring/VaultStorage.test.ts

```diff
@@ -0,0 +1,318 @@
+
+import { describe, expect, it, vi } from 'vitest';
+import Browser from 'webextension-polyfill';
+
+import { VaultStorage } from './VaultStorage';
+import { SESSION_STORAGE } from './storage-utils';
+import * as Bip39 from '_shared/utils/bip39';
+import {
+    testEd25519Serialized,
+    testVault,
+    testVault1,
+} from '_src/test-utils/vault';
+
+describe('VaultStorage', () => {
+    describe('create', () => {
+        it('throws if already initialized', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({
+                vault: {},
+            });
+            await expect(vaultStorage.create('12345')).rejects.toThrow(
+                'Mnemonic already exists, creating a new one will override it. Clear the existing one first.'
+            );
+        });
+
+        it('uses random entropy when not provided and creates the vault', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({});
+            const storageSet = vi
+                .spyOn(Browser.storage.local, 'set')
+                .mockResolvedValue();
+            const getRandomEntropySpy = vi
+                .spyOn(Bip39, 'getRandomEntropy')
+                .mockReturnValue(new Uint8Array(32));
+            await vaultStorage.create('12345');
+            expect(storageSet).toHaveBeenCalledOnce();
+            expect(storageSet.mock.calls[0][0]).toMatchObject({
+                vault: {
+                    v: 2,
+                    data: expect.stringContaining('data'),
+                },
+            });
+            expect(getRandomEntropySpy).toHaveBeenCalledOnce();
+        });
+
+        it('uses the provided entropy and creates the vault', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({});
+            const storageSet = vi
+                .spyOn(Browser.storage.local, 'set')
+                .mockResolvedValue();
+            const getRandomEntropySpy = vi
+                .spyOn(Bip39, 'getRandomEntropy')
+                .mockReturnValue(new Uint8Array(32));
+            await vaultStorage.create(
+                '12345',
+                '842a27e29319123892f9ba8d9991c525'
+            );
+            expect(storageSet).toHaveBeenCalledOnce();
+            expect(storageSet.mock.calls[0][0]).toMatchObject({
+                vault: {
+                    v: 2,
+                    data: expect.stringContaining('data'),
+                },
+            });
+            expect(getRandomEntropySpy).not.toHaveBeenCalled();
+        });
+    });
+
+    describe('unlock', () => {
+        it('throws if wallet is not initialized', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({});
```

Comment:
The more I see this the more I'm convinced we need to figure out a better way to mock this. Maybe the constructor can take in a storage adapter, and defaults to `Browser.storage.local`, but we can provide our own pre-set one in tests or something like that? But I think we def need to improve this before merging.